File: csv/check_links.py
"""
Check validity of URLs in CSV files.
This script should be run before building the new catalog.
It results in a CSV output file ('csv/link_report.csv') containing broken links.
"""

import logging
import sys
from pathlib import Path
import pandas as pd
import urllib.request
import urllib.error
import ssl

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
logger = logging.getLogger("link_checker")

def check_url(url: str) -> tuple[bool, str | int]:
    """Check if a URL is reachable."""
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    # Create SSL context that ignores certificate errors
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE

    try:
        # 1. Try HEAD first
        req = urllib.request.Request(url, headers=headers, method='HEAD')
        with urllib.request.urlopen(req, timeout=15, context=ctx) as response:
            return True, response.status
    except Exception:
        try:
            # 2. Fallback to GET with Range (to avoid downloading large files)
            get_headers = headers.copy()
            get_headers['Range'] = 'bytes=0-10'
            req = urllib.request.Request(url, headers=get_headers, method='GET')
            with urllib.request.urlopen(req, timeout=15, context=ctx) as response:
                return True, response.status
        except Exception:
            try:
                # 3. Fallback to GET without Range (some servers reject Range)
                req = urllib.request.Request(url, headers=headers, method='GET')
                with urllib.request.urlopen(req, timeout=15, context=ctx) as response:
                    return True, response.status
            except urllib.error.HTTPError as e:
                # Accept 401/403 as valid (resource exists but is protected/login required)
                if e.code in [401, 403]:
                    return True, e.code
                return False, e.code
            except Exception as e:
                return False, str(e)

def check_links(csv_paths: list[str]):
    """Check links in the provided CSV files."""
    has_errors = False
    link_report = []
    
    for path in csv_paths:
        file_path = Path(path)
        if not file_path.exists():
            logger.error(f"File not found: {file_path}")
            continue
            
        logger.info(f"Checking file: {file_path}")
        try:
            df = pd.read_csv(file_path)
        except Exception as e:
            logger.error(f"Failed to read {file_path}: {e}")
            continue
        
        if 'asset_links' not in df.columns:
            logger.warning(f"'asset_links' column missing in {file_path}")
            continue
            
        for idx, row in df.iterrows():
            asset_links = row['asset_links']
            
            # Skip if empty or NaN
            if pd.isna(asset_links) or str(asset_links).strip() == "":
                continue
                
            # Handle space separator
            links = str(asset_links).split()
            
            for link in links:
                if not link:
                    continue
                    
                success, status = check_url(link)
                item_title = row.get('title_item', f'Row {idx + 2}')
                data_overview = row.get('data_overview_link', None)

                if success:
                    continue

                has_errors = True
                logger.error(f"Broken link in '{item_title}': {link} (Status: {status})")
                link_report.append({
                    'file': str(path),
                    'row_number': idx + 2,
                    'title_item': item_title,
                    'data_overview_link': data_overview,
                    'link': link,
                    'status': str(status),
                    'success': success
                    })

    num_broken_links = sum(not item['success'] for item in link_report)
    if num_broken_links > 0:
        pd.DataFrame(link_report).to_csv("csv/link_report.csv", index=False)
        logger.info("Full link report saved to 'csv/link_report.csv'.")
        logger.info(f"Number of broken links: {num_broken_links}")
    else:
        logger.info("All CSVs are valid: no broken links found.")

    if has_errors:
        sys.exit(1)

if __name__ == "__main__":
    # Default paths based on project structure 
    paths = ['csv/hazard.csv', 'csv/expvul.csv']
    check_links(paths)

File: csv/check_links.R
#!/usr/bin/env Rscript

# Check validity of URLs in CSV files.

# Check for required package
if (!requireNamespace("httr", quietly = TRUE)) {
  stop("Package 'httr' is needed for this script to work. Please install it with install.packages('httr').", call. = FALSE)
}

check_url <- function(url) {
  # User agent similar to the Python script
  ua <- httr::user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36")
  
  tryCatch({
    # Try HEAD first to save bandwidth
    response <- httr::HEAD(url, ua, httr::timeout(10))
    status <- httr::status_code(response)
    
    if (status >= 400) {
      # Fallback to GET if HEAD fails (some servers block HEAD)
      response <- httr::GET(url, ua, httr::add_headers(Range = "bytes=0-10"), httr::timeout(10))
      status <- httr::status_code(response)
    }
    
    if (status < 400) {
      return(list(success = TRUE, status = status))
    } else {
      return(list(success = FALSE, status = status))
    }
  }, error = function(e) {
    return(list(success = FALSE, status = e$message))
  })
}

check_links <- function(csv_paths) {
  has_errors <- FALSE
  # Initialize data frame to store link details
  link_report <- data.frame(
    file = character(),
    row_number = integer(),
    title_item = character(),
    link = character(),
    status = character(),
    success = logical(),
    stringsAsFactors = FALSE
  )
  
  for (path in csv_paths) {
    if (!file.exists(path)) {
      message(sprintf("File not found: %s", path))
      next
    }
    
    message(sprintf("Checking file: %s", path))
    
    # Read CSV
    # Using base R read.csv to avoid extra dependencies like readr/tidyverse if possible,
    # but ensuring strings are not factors.
    tryCatch({
      df <- read.csv(path, stringsAsFactors = FALSE, encoding = "UTF-8")
    }, error = function(e) {
      message(sprintf("Failed to read %s: %s", path, e$message))
      return(NULL) 
    })
    
    if (is.null(df)) next
    
    if (!"asset_links" %in% names(df)) {
      message(sprintf("'asset_links' column missing in %s", path))
      next
    }
    
    # Iterate through rows
    for (i in seq_len(nrow(df))) {
      asset_links <- df$asset_links[i]
      
      # Skip if empty or NA
      if (is.na(asset_links) || trimws(asset_links) == "") {
        next
      }
      
      # Handle space separator
      links <- strsplit(asset_links, " ")[[1]]
      
      for (link in links) {
        link <- trimws(link)
        if (link == "") next
        
        result <- check_url(link)
        item_title <- if ("title_item" %in% names(df)) df$title_item[i] else sprintf("Row %d", i + 1)
        
        if (!result$success) {
          has_errors <- TRUE
          message(sprintf("Broken link in '%s': %s (Status: %s)", item_title, link, result$status))
        }
        
        # Add to report
        link_report <- rbind(link_report, data.frame(
          file = path,
          row_number = i + 1, # Assuming header row exists
          title_item = item_title,
          link = link,
          status = as.character(result$status),
          success = result$success,
          stringsAsFactors = FALSE
        ))
      }
    }
  }
  
  write.csv(link_report, "link_report.csv", row.names = FALSE)
  message("Full link report saved to 'link_report.csv'.")

  if (has_errors) {
    quit(status = 1)
  } else {
    message("All links checked successfully.")
  }
}

# Main execution
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
  # Default paths based on project structure
  paths <- c("csv/hazard.csv", "csv/expvul.csv")
} else {
  paths <- args
}

check_links(paths)
